### DISCUSSION TITLE
Struggling to get ESPresense to work with M5Stamp Pico

### DISCUSSION THREAD

**Discussion by MilesAheadToo** [ESPresense/ESPresense#459] [Support]
Let me start by saying that I am far from an expert in ESP32, BLE, or even HA. But the project looks like exactly what I need to automate opening the front gate and turning on/off the alarm when people leave/return to the house. 

I have been trying all day to get it going but cannot see what it is that I am doing incorrectly.

I have bought 3 M5Stamp Picos that I am now trying to install and integrate with HA. I will only step through one M5Stamp setup as I have not even tried the others.

This is the configuration of the M5Stamp.
![image](https://user-images.githubusercontent.com/78782836/165993138-6d2e1814-8a52-4edf-8f40-4c36a70b667a.png)![image](https://user-images.githubusercontent.com/78782836/165992893-4e192dd1-09c2-4c2a-b284-896fa83bbb93.png)
![image](https://user-images.githubusercontent.com/78782836/165992961-f6cc132f-7ee0-4598-8f90-407fb9cb71ff.png)

I started with the defaults for the known MAC addresses and included devices blank and got the same results. I also started with the default less than and greater than distances.

This is the ESPresense log file with 1 restart.

`Connecting to WiFi SSID 'Miles-Home'.....................192.xxx.xxx.40
Firmware:     esp32
Version:      v2.1.1
IP address:   192.xxx.xxx.40
DNS address:  192.xxx.xxx.1
Hostname:     espresense-gate
Room:         Gate
MQTT server:  192.xxx.xxx.36:1883
Max Distance: 50.00
PIR Sensor:   disabled
Radar Sensor: disabled
Query:        
Include:      
Exclude:      
Known Macs:   7c2f80c39b11 7c2f80c3a111 7c2f80c39bbb 7c2f80c36327 7c2f80c36e05 e3e1323cc0ab e44b5efe1d0e 
Count Ids:    known:7c2f80c39b11 known:7c2f80c3a111 known:7c2f80c39bbb known:7c2f80c36327 known:7c2f80c36e05 tile:e3e1323cc0ab tile:e44b5efe1d0e 
1 New   | MAC: 498909c69d51, ID: md:00e0:8                                                    
1 New   | MAC: 4cc95e068c4f, ID: 4cc95e068c4f                                                 
1 New   | MAC: d2d2d3bf686c, ID: sonos:d2d2d3bf686c                                           
1 New   | MAC: e3e1323cc0ab, ID: known:e3e1323cc0ab                                           
1 New   | MAC: 7869819da8cf, ID: exp:20                                                       eba5004dc61ad39c17e8a790659bea11a7027225
1 New   | MAC: 263ec64a3101, ID: msft:cdp:0902                                                304fe94220b9e95bef6361985a29acdddbd32224
1 New   | MAC: 4d3e8769ab23, ID: md:00e0:8                                                    
1 Close | MAC: e44b5efe1d0e, ID: known:e44b5efe1d0e                                           (0.00m) -33dBm
1 New   | MAC: e44b5efe1d0e, ID: known:e44b5efe1d0e                                           
1 Close | MAC: 7c2f80c39bbb, ID: known:7c2f80c39bbb                                           (0.00m) -39dBm
1 New   | MAC: 7c2f80c39bbb, ID: known:7c2f80c39bbb                                           
1 New   | MAC: 7c2f80c3a1ee, ID: 7c2f80c3a1ee                                                 
1 New   | MAC: 630b3fec35de, ID: md:00e0:8                                                    
1 New   | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           
1 C# +1 | MAC: 7c2f80c39bbb, ID: known:7c2f80c39bbb                                           (0.10m) 2940ms
1 C# +1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (0.00m) 26989ms
1 C# -1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (0.00m) 30001ms
1 C# +1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (4.54m) 3ms
1 C# -1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (5.41m) 30001ms
1 C# +1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (5.27m) 4ms
1 C# -1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (5.14m) 30001ms
1 C# +1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (4.98m) 3ms
1 C# -1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (4.98m) 30001ms
Not updating from (sc=302): https://github.com/ESPresense/ESPresense/releases/download/v2.1.1/esp32.bin
1 C# +1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (4.89m) 3ms
1 New   | MAC: 4f70af91bd45, ID: md:00e0:8                                                    
1 C# -1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (5.15m) 30001ms
1 C# +1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (5.15m) 4ms
1 New   | MAC: 7f42fc7198ea, ID: md:00e0:8                                                    
1 New   | MAC: 6409690a3a19, ID: exp:20                                                       45123dd043446b0120c6bea83dd038d8d63953e0
1 New   | MAC: 1cbb00c62e37, ID: msft:cdp:0902                                                d41efe664d03f7d17ca846e43d5cd6cd5fdb069b
1 C# -1 | MAC: 7c2f80c39b11, ID: known:7c2f80c39b11                                           (5.61m) 30001ms
[E][main.cpp:291] onMqttDisconnect(): Disconnected from MQTT; reason 0

ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:8896
load:0x40080400,len:5828
entry 0x400806ac
Setup network
Connecting to WiFi SSID 'Miles-Home'......................192.xxx.xxx.40
Firmware:     esp32
Version:      v2.1.1
IP address:   192.xxx.xxx.40
DNS address:  192.xxx.xxx.1
Hostname:     espresense-gate
Room:         Gate
MQTT server:  192.xxx.xxx.36:1883
Max Distance: 50.00
PIR Sensor:   disabled
Radar Sensor: disabled
Query:        
Include:      
Exclude:      
Known Macs:   7c2f80c39b11 7c2f80c3a111 7c2f80c39bbb 7c2f80c36327 7c2f80c36e05 e3e1323cc0ab e44b5efe1d0e 
Count Ids:    known:7c2f80c39b11 known:7c2f80c3a111 known:7c2f80c39bbb known:7c2f80c36327 known:7c2f80c36e05 tile:e3e1323cc0ab tile:e44b5efe1d0e 
`
Everything looks fine to me.

 And _**MQTT Explorer**_ is reporting results that I would expect.
![image](https://user-images.githubusercontent.com/78782836/165995866-2e9cb67d-8dd2-4c0f-b800-fc6a671f2e33.png)

![image](https://user-images.githubusercontent.com/78782836/165995518-f29801b1-8a16-4958-b3e1-8f040509206f.png)

But I cannot see any of the sensors in HA.  This is my Mosquitto configuration:
![image](https://user-images.githubusercontent.com/78782836/165996255-d55f1073-7ccf-4263-9337-cb99fd00b03d.png)

On restarting Mosquitto to get the following log, which looks fine to me.

`[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] mosquitto.sh: executing... 
[18:51:16] INFO: Setting up user mosquitto
[18:51:17] INFO: Certificates found: SSL is available
[cont-init.d] mosquitto.sh: exited 0.
[cont-init.d] nginx.sh: executing... 
[cont-init.d] nginx.sh: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[18:51:17] INFO: Starting NGINX for authentication handling...
[18:51:17] INFO: Starting mosquitto MQTT broker...
1651254677: mosquitto version 1.6.12 starting
1651254677: |-- *** auth-plug: startup
[18:51:17] INFO: Successfully send discovery information to Home Assistant.
[18:51:17] INFO: Successfully send service information to the Supervisor.
1651254677: Config loaded from /etc/mosquitto/mosquitto.conf.
1651254677: Loading plugin: /usr/share/mosquitto/auth-plug.so
1651254677:  ├── Username/password checking enabled.
1651254677:  ├── TLS-PSK checking enabled.
1651254677:  └── Extended authentication not enabled.
1651254677: Opening ipv4 listen socket on port 1883.
1651254677: Opening ipv6 listen socket on port 1883.
1651254677: Opening websockets listen socket on port 1884.
1651254677: Opening ipv4 listen socket on port 8883.
1651254677: Opening ipv6 listen socket on port 8883.
1651254677: Opening websockets listen socket on port 8884.
1651254677: Warning: Mosquitto should not be run as root/administrator.
1651254677: mosquitto version 1.6.12 running
1651254677: New connection from 127.0.0.1 on port 1883.
1651254677: Socket error on client <unknown>, disconnecting.
1651254678: New connection from 172.30.32.1 on port 1883.
1651254678: New client connected from 172.30.32.1 as 4b4FoFOADcgXXcdyI8TBsd (p2, c1, k60, u'mosquitto').`

This is what the MQTT integration looks like.
![image](https://user-images.githubusercontent.com/78782836/165998502-9311eb53-ca43-48fd-b4e9-2f8d73cbd479.png)

I define the different beacons in a `sensors.yaml` file that I reference in the overall `configuration.yaml` by `sensor: !include sensors.yaml`

`- platform: mqtt_room
  device_id: "7c2f80c36327"
  name: 'Keys - Cleaner'
  state_topic: 'espresense/rooms'
  timeout: 60
  away_timeout: 120 # number of seconds after which the enitity will get status not_home

- platform: mqtt_room
  device_id: "7c2f80c46e05"
  name: 'Keys - Britta'
  state_topic: 'espresense/rooms'
  timeout: 60
  away_timeout: 120 # number of seconds after which the enitity will get status not_home

- platform: mqtt_room
  device_id: "7c2f80c39b11"
  name: 'Keys - Spare'
  state_topic: 'espresense/rooms'
  timeout: 60
  away_timeout: 120 # number of seconds after which the enitity will get status not_home

- platform: mqtt_room
  device_id: "7c2f80c3a111"
  name: 'Keys - Trevor'
  state_topic: 'espresense/rooms'
  timeout: 60
  away_timeout: 120 # number of seconds after which the enitity will get status not_home

- platform: mqtt_room
  device_id: "7c2f80c39bbb"
  name: 'Car - Fiat'
  state_topic: 'espresense/rooms'
  timeout: 60
  away_timeout: 120 # number of seconds after which the enitity will get status not_home

- platform: mqtt_room
  device_id: "tile:e3e1323cc0ab"
  name: 'Car - MB250GLC'
  state_topic: 'espresense/rooms'
  timeout: 60
  away_timeout: 120 # number of seconds after which the enitity will get status not_home
`
But when I look at the entities in HA, they are not operating correctly.
![image](https://user-images.githubusercontent.com/78782836/165999189-d08753fe-16c5-4187-abc7-44d17d21e726.png)

Below is the diagnostic run from within HA for the `ESPresense Gate` device.
[mqtt-893b5dcc7bd3c9150db49010c49d0ed6-ESPresense Gate-5996b97c6d0c33545da67453b55aa4b5.json (1).txt](https://github.com/ESPresense/ESPresense/files/8592806/mqtt-893b5dcc7bd3c9150db49010c49d0ed6-ESPresense.Gate-5996b97c6d0c33545da67453b55aa4b5.json.1.txt)

**Comment by DTTerastar** [comment]
The ESP32 can't do SSL to MQTT.  I think you have that enabled.

**Reply by MilesAheadToo** [reply]
Thanks for the reply @DTTerastar. Unfortunately, I need more information. 

I do not recall enabling SSL. Do you think I enabled it on HA MQTT? I can see the M5Stamp Pico in HA through the MQTT add-in. All the sensors seem to be returning values, so I presume ESPresense is communicating with HA with no problems.
![image](https://user-images.githubusercontent.com/78782836/166106129-a91dc15a-0570-4aca-8f30-50c349c97e60.png)

When I click on `Visit Device` it takes me to the M5Stamp Pico website for settings.

Is there anything in the attached diagnostics that would indicate that the M5Stamp pico is configured incorrectly or not operating as expected?
[mqtt-893b5dcc7bd3c9150db49010c49d0ed6-ESPresense Gate-5996b97c6d0c33545da67453b55aa4b5.json (2).txt](https://github.com/ESPresense/ESPresense/files/8596919/mqtt-893b5dcc7bd3c9150db49010c49d0ed6-ESPresense.Gate-5996b97c6d0c33545da67453b55aa4b5.json.2.txt)

It's just that I cannot see the beacons.

**Comment by DTTerastar** [comment]
Oh, I'm sorry, I saw the certificate and jumped to conclusions.  It looks like you're missing the known: prefix in the hass config.  Hass documentation assumes all ids are just macs, but ESPresense has various types of ids.  You need the entire value in the hass config.

**Reply by MilesAheadToo** [reply]
Seems like this was the problem. I deleted both known macs and the IDs in the M5Stamp Pico configuration and everything seems to be ok.

Thanks for the help.

**Reply by DTTerastar** [reply]
It's fine to make them known just put

```platform: mqtt_room
device_id: "known:7c2f80c39bbb"
name: 'Car - Fiat'
state_topic: 'espresense/rooms'
timeout: 60
away_timeout: 120 # number of seconds after which the enitity will get status not_home
```

**Reply by MilesAheadToo** [reply]
Thank you @DTTerastar. I'll get comfortable first, and then try with known.

Is there any way to get distance reported? What I am trying to do is have 3 sensors along my driveway. 

If the beacon is 'not_home' and then appears at the 'Gate' room I know that that car or person is approaching the gate and I can open it. Similarly, once they get near the front door I can disarm the alarm.

On the other hand, if one of the beacons is in a room other than 'Gate', in other words somewhere at 'home', and then pops into Gate I can open the gate to let them out and arm the alarm if all the beacons are 'not_home'.

But I would also like to know if there is movement. Ideally, I would like to get all the distances for a beacon. You must be doing this already to determine the room in which the beacon is located. But I can't see how to expose this information in ESPresense.

**Reply by DTTerastar** [reply]
Yes, the distance is a property of the json so you can use just a generic mqtt sensor to get it:

```
- platform: mqtt
  state_topic: espresense/devices/apple:watch6-1/dining
  value_template: '{{ value_json.distance }}'
  unit_of_measurement: 'm'
  name: 'Darrell Watch Dining Proximity'
```

**Reply by MilesAheadToo** [reply]
I presume this is the distance to the closest sensor?

**Reply by DTTerastar** [reply]
No dining is a particular base station.  So you'd put 3 of these in for the 3 along the driveway.

**Reply by MilesAheadToo** [reply]
Got it. 

I was referring to the attribute returned by, for example,` {{ state_attr('sensor.car_fiat','distance') }}`. I presume this is the distance from the base station returned by `{{ states('sensor.car_fiat') }}` as the room?

I have 3 base stations and 8 beacons, so that is 24 sensors. Is it possible to retrieve the distance of a beacon from all the base stations as a set of paired attributes?

**Comment by MilesAheadToo** [comment]
Much appreciated Darrel. You could have replied with RTFM!! I just saw it in the documentation.

Sent: 30 April 2022 17:07
To: ESPresense/ESPresense ***@***.***>
Cc: MilesAheadToo ***@***.***>; Author ***@***.***>
Subject: Re: [ESPresense/ESPresense] Struggling to get ESPresense to work with M5Stamp Pico (Discussion #459)

 

Yes, the distance is a property of the json so you can use just a generic mqtt sensor to get it:

- platform: mqtt
  state_topic: espresense/devices/apple:watch6-1/dining
  value_template: '{{ value_json.distance }}'
  unit_of_measurement: 'm'
  name: 'Darrell Watch Dining Proximity'
